graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,175 @@
+
+@dataclass
+class PaginationAdvisory(ABC):
+    message: str = field(init=False)
+
+
+@dataclass
+class PaginationFieldNotSpecified(PaginationAdvisory):
+    vertex_name: str
+
+    def __post_init__(self):
+        """Initialize a human-readable message."""
+        self.message = (
+            "Specifying a pagination field for vertex {} in the QueryPlanningSchemaInfo "
+            "would have made this vertex eligible for pagination, and enabled a better "
+            "pagination plan.".format(self.vertex_name)
+        )
+
+
+@dataclass
+class InsufficientQuantiles(PaginationAdvisory):
+    vertex_name: str
+    field_name: str
+    current_resolution: int
+    desired_resolution: int
+
+    def __post_init__(self):
+        """Initialize a human-readable message."""
+        self.message = (
+            "Pagination would have been more successful if more quantiles were provided "
+            "for {}.{}. Currently there are {}, ideally there should be {}".format(
+                self.vertex_name, self.field_name, self.current_resolution, self.desired_resolution
+            )
+        )
+
 
+@dataclass(frozen=True)
+class VertexPartitionPlan:
+    """Plan to split the query at a certain vertex into a certain number of pages."""
 
-# The intent to split the query at a certain vertex into a certain number of pages.
-VertexPartition = namedtuple(
-    "VertexPartition",
-    (
-        "query_path",  # Tuple[field name : str] leading to the vertex to be split
-        "number_of_splits",  # The number of subdivisions intended for this vertex
-    ),
-)
+    # field names leading to the vertex to be split
+    query_path: Tuple[str, ...]
 
+    # field to use for pagination
+    pagination_field: str
 
-# The intent to split the query with a combination of VertexPartitions
-PaginationPlan = namedtuple("PaginationPlan", ("vertex_partitions",))  # List[VertexPartition]
+    # The number of subdivisions intended for this vertex
+    number_of_splits: int
 
 
-# TODO(bojanserafimov): Make this function return a best effort pagination plan
-#                       when a good one is not found instead of returning None.
-def try_get_pagination_plan(schema_info, query_ast, number_of_pages):
-    """Make a PaginationPlan for the given query and number of desired pages if possible."""
+@dataclass(frozen=True)
+class PaginationPlan:
+    """Plan to split the query with a combination of VertexPartitionPlans."""
+
+    vertex_partitions: Tuple[VertexPartitionPlan, ...]
+
+
+def get_num_pages_generated_by_plan(plan: PaginationPlan) -> int:
+    """Return the number of pages that a PaginationPlan would generate."""
+    number_of_pages = 0
+    for vertex_partition in plan.vertex_partitions:
+        number_of_pages *= vertex_partition.number_of_splits
+    return number_of_pages
+
+
+def get_pagination_plan(
+    schema_info: QueryPlanningSchemaInfo, query_ast: DocumentNode, number_of_pages: int
+) -> Tuple[PaginationPlan, List[PaginationAdvisory]]:
```

Comment:
Consider changing `List[PaginationAdvisory]` to `Tuple[PaginationAdvisory, ...]` since we don't really need it to be mutable / a list at all.